### DISCUSSION TITLE
[Authentication service] REDIS_HOSTS

### DISCUSSION THREAD

**Discussion by georgev-97** [mia-platform/community#360] [Q&A] [answered]
### Which section?

Console

### Which is the version?

12

### Description

Hello, I am trying to configure the authentication service by following the documentation but I can't get it to connect to my local redis nor to a cloud redis.
The local redis I've tried is a microservice built with the docker-image _docker.io/bitnami/redis:6.2.7-debian-11-r3_ exposed on the 6379 port with no password.
What should I set into the REDIS_HOSTS to get it to connect to my local redis instance?
I've also tried to configure a cloud redis by setting REDIS_HOSTS as `redis://username:password@ipaddress:port/dbnum` with the same error:
`{"error":"redis error","level":60,"msg":"redis connection failed","time":1703756628499}`

P.S. is it possible to use the authentication service to implement the "username/password flow" instead of the "authorization code flow"?

**Comment by malta895** [comment, marked as answer]
Hi @georgev-97,
you probably spotted an oversight in the documentation, the correct way to connect to Redis is the following:
- set the `REDIS_HOSTS`  env in the form `ipaddress:port`. Protocol, credentials and `dbnum` are not needed.
- `REDIS_USERNAME` and `REDIS_PASSWORD` are indeed required if your server requires authentication, and they are not deprecated as erroneously stated in the docs. Please set them accordingly.

In the meantime we will take care of fixing the documentation, thank you!

Regarding your second question, yes, the authentication service supports the ROPC flow, but please consider that the support is partial and deprecated, and only a few providers support it.

The reason is that the flow is discouraged as it is considered legacy [by the specification](https://datatracker.ietf.org/doc/html/rfc6749#section-10.7).

Nevertheless, if you decide to use it, you can `POST /oauth/token` API with the following body:
```json
{
    "grant_type": "password",
    "username": "yourusername",
    "password": "yourpassword",
    "appId": "yourAppId",
    "providerId": "yourProviderId"
}
```

Given the correct credentials, the API will respond `200 OK`:
```json
{
    "accessToken": "my-access-token",
    "refreshToken": "my-refresh-token",
    "expireAt": 1234567890
}
```

**Reply by malta895** [reply]
Please note that in the `oauth/token` request body you also need to add a `providerId` and an `appId`, corresponding to the relevant configuration of your configurations file. I updated the response accordingly, sorry for the oversight!

**Reply by georgev-97** [reply]
Thank you for you answer!
I replaced the REDIS_HOSTS with `ipaddress:port` as suggested but now I get a different error which may not be related to redis:
`{"level":60,"msg":"fails to create client: baseURL should be an absolute url","time":1703770952938}`

**Reply by malta895** [reply]
As answered in #361, the error is caused by a malformed `USERS_CRUD_BASE_URL`.
